### PR TITLE
Upgrades the space heater

### PR DESCRIPTION
**Pull request by YotaXP**
You guys remember those black cubey things in maint that supposedly make things less cold?  I'm sure most people don't.
## Contents of this PR:
- New Nano UI interface.
- The heater can now double as a cooling device! (My own animation included!)
- The heater can now be built and deconstructed like any other machine.  (1x laser, 1x capacitor, 3x wire)
- New circuit board for the heater. (`origin_tech = "programming=2;engineering=2"`)
- Heating range altered. (Formerly 20°C to 90°C) See below:
### Upgradability

| Component | Function | Tier 1 | Tier 4 |
| --- | --- | --- | --- |
| Laser | Determines rate of heating/cooling. | 1x Speed | 4x Speed |
| Capacitor | Efficiency, and the heating/cooling range. | 1x Power Draw<br>0°C to 60°C | 2/5x Power Draw<br>-90°C to 150°C |
### Pics or it didn't happen

Icon Preview
Off / Standby (New) / Heat / Cool (New)
NanoUI Closed
NanoUI Open

:cl:
rscadd: Rolled out a new line of space heaters around the station.  This new model has the functionality to cool down the environment, as well as heat it.  Additionally, they can be built and upgraded like any other machine.
/:cl:

### CODE REVIEW

**Comment by YotaXP** [comment]
Please suggest new names, new descriptions, and new numbers, because I clearly don't know the first thing about balancing.

**Comment by Razharas** [comment]
Names overlap with atmos heater/cooler
Maybe rename it into air conditioner or something since it can cool now too

**Comment by Razharas** [comment]
Also remove the debug

File: code/game/machinery/spaceheater.dm [hunk 1]
```diff
@@ -60,128 +131,142 @@
...
-				usr.visible_message("[usr] removes the power cell from \the [src].", "<span class='notice'>You remove the power cell from \the [src].</span>")
+		var/minTemp = settableTemperatureMedian - settableTemperatureRange
+		var/maxTemp = settableTemperatureMedian + settableTemperatureRange
+		targetTemperature = dd_range(max(1, minTemp), maxTemp, round(value, 1))
+		SSnano.update_uis(src)
 
+	else if(href_list["cellremove"] && panel_open)
+		if(cell)
+			if(usr.get_active_hand())
+				user << "<span class='warning'>You need an empty hand to remove \the [cell]!</span>"
```
Comment (Razharas) [inline code/game/machinery/spaceheater.dm]: user => usr

**Comment by Supermichael777** [comment]
t4 is not nearly retarded enough. where are my negative kelven temperatures?

File: code/game/machinery/spaceheater.dm [hunk 2]
```diff
@@ -60,128 +129,142 @@
...
-		onclose(user, "spaceheater")
-
-
+	if(!user)
+		return
+	interact(user)
 
+/obj/machinery/space_heater/attack_paw(mob/user)
+	if(!user)
+		return
```
Comment (phil235) [inline code/game/machinery/spaceheater.dm]: Are those two `if(!user)` check necessary? If some code actually manages to call attack_hand with a null user it'd be a bug and we shouldn't try to hide it.

**Comment by duncathan** [comment]
@Supermichael777 atmos coolers used to be able to go to negative temperatures at tier 4. It fucked everything up real bad and had to be patched out. 

That being said, there's a define (TCMB) that is the temperature of space files. That's a fun idea for t4.

**Comment by as334** [comment]
Yeah we're not supposed to make things colder than 2.7 kelvin. It causes problems or something.

**Comment by Alek2ander** [comment]
Make the button corresponfing to the currently active mode green (selected) instead of gray (inactive)

**Comment by neersighted** [comment]
@YotaXP Also please add icons to the power button

File: code/game/machinery/spaceheater.dm [hunk 3]
```diff
@@ -1,40 +1,84 @@
...
-	var/on = 0
-	var/open = 0
-	var/set_temperature = 50		// in celcius, add T0C for kelvin
-	var/heating_power = 40000
+	var/on = FALSE
+	var/mode = HEATER_MODE_STANDBY
+	var/setMode = "auto" // Anything other than "heat" or "cool" is considered auto.
+	var/targetTemperature = T20C
+	var/heatingPower = 40000
+	var/effeciency = 20000
```
Comment (phil235) [inline code/game/machinery/spaceheater.dm]: "efficiency"

**Comment by neersighted** [comment]
Also, if I'm not mistaken, space_heater.tmpl is missing from this PR.

**Comment by YotaXP** [comment]
All suggestions except for the T4 one have been implemented.  Should I make it more retarded?  Currently, both extremes are able to deal minor heat/cold damage.

Should it be emaggable, and if so, what should it do?

**Comment by swankcookie** [comment]
for Emagged behaviors, maybe something like touching it will deal burn damage (so you can be pushed into it for burn. I'd be wary of making it too OP though, seeing as heaters are currently placed every 5 feet.

**Comment by Zilenan91** [comment]
I'd love nothing better than to burn people alive with space heaters.


Emagging should make them explode into fireballs when turned on, with the ability to use a signaller to trigger it manually.

**Comment by swankcookie** [comment]
I thought about the possibility of making them explode, but as i said earlier, given how flush the maps are with them, it hardly seems like a fair advantage.

**Comment by bear1ake** [comment]
Like it :+1:

**Comment by Supermichael777** [comment]
just make emaging force it to heat to dangerous tempatures and ignite plasma.

**Comment by swankcookie** [comment]
is their a graph somewhere of the temps needed too light plasma against its ratio to the surrounding air?

**Comment by YotaXP** [comment]
How about when emagged: The 'auto' mode is disabled.  In 'heat' mode, the target temp gets locked at 999.  In 'cool' mode, the target temp gets locked at 3.

???

**Comment by Cheridan** [comment]
that sounds ok. is that something you want to do here, or i can merge this and you can pr that later since this otherwise looks good to go?

**Comment by YotaXP** [comment]
Might as well merge it.

**Pull request merged**